### PR TITLE
Hooks

### PR DESCRIPTION
**Pull request by alexfedoseev**

### CODE REVIEW

**Comment by thitemple** [comment]
Thanks for this!! I just updated my app to use it, and I think it's looking great. I love the new api instead of having to use a component to do the same thing.

**Comment by alexfedoseev** [comment]
Agreed 😃

File: examples/App.re [hunk 1]
```diff
@@ -9,69 +7,79 @@ type state = {route};
...
+      {route: initialRoute},
+    );
 
-let make = _ => {
-  ...component,
-  initialState: () => {route: getInitialRoute()},
-  reducer: (action, _) =>
-    switch (action) {
-    | UpdateRoute(route) => React.Update({route: route})
+  React.useEffect1(
```
Comment (tcoopman) [inline examples/App.re]: you could use `React.useEffect0` here (https://github.com/reasonml/reason-react/blob/bb480418d99acc49bb5df5df512f68646cce26ad/src/React.re#L131) this sends an empty array as argument.

File: examples/LoginForm.re [hunk 2]
```diff
@@ -56,15 +56,20 @@ module LoginForm = {
...
-let make = _ => {
-  ...component,
-  render: _ =>
-    <LoginFormContainer
-      initialState={email: "", password: "", rememberMe: false}
-      onSubmit={(state, form) => {
+[@react.component]
+let make = () => {
+  let initialState =
+    React.useMemo1(
```
Comment (tcoopman) [inline examples/LoginForm.re]: Is this use of useMemo useful here? Why not extract initial state outside of the component?

File: examples/App.re [hunk 3]
```diff
@@ -9,69 +7,79 @@ type state = {route};
...
+    () => {
+      let watcherID =
+        ReasonReactRouter.watchUrl(url =>
```
Comment (rickyvetter) [inline examples/App.re]: Why not `ReasonReactRouter.useUrl`? There is a small issue with this implementation in concurrent mode which is that this effect is async. If the url changes between mount and the effect you won't detect it. `useUrl` accounts for this case - and also removes a lot of boilerplate :)

File: examples/LoginForm.re (hunk 2 above)
Comment (rickyvetter) [reply examples/LoginForm.re]: If you're doing this to make `initialState` private without adding an rei we support the block syntax for components. Also `useMemo0` is equivalent to `useMemo1` with an empty array.

```reason
[@react.component]
let make = {
  let initializationStuff = 1;
  (~prop) => ...
};
```

File: examples/App.re (hunk 1 above)
Comment (alexfedoseev) [reply examples/App.re]: TIL, will do 👍

File: examples/App.re (hunk 3 above)
Comment (alexfedoseev) [reply examples/App.re]: TIL as well, will use it 👍

File: examples/LoginForm.re (hunk 2 above)
Comment (alexfedoseev) [reply examples/LoginForm.re]: Ohh very nice! TBH I used it this way for no particular reason, just getting myself familiar with hooks. Will address, thanks for suggestions!

**Pull request merged**